### PR TITLE
include: Prefix includes to use a scope

### PR DESCRIPTION
**Pull request by yperess**
Move include paths and add new target_include_directories to support
backwards compatibility:
* /include -> /include/zephyr
  example: <irq.h> -> <zephyr/irq.h>
* /kernel/include -> /kernel/include/zephyr/kernel/
  example: <thread.h> -> <zephyr/kernel/threads.h>

Issue #41543

Signed-off-by: Yuval Peress <peress@google.com>

### CODE REVIEW

**Review by andyross — DISMISSED** [review]
Ooooph....  standard Grumpy Curmudgeon complaints apply here: touching every file in the project makes patch rebasing across versions a pain, affects users with portability concerns, and generally just leads to headache.

That said, yes, our include structure is a mess.   Yes, it's probably overdue for a cleanup.  And yes, I like this scheme better than what we have (rather, "the most charitable interpretation of the design intent of what we have", heh).  So I guess this is +1 from me.

But can we please make sure this is what we really want and that we're sure we're not just going to need to do this again next bikeshed season?  I still bear scars from our "int -> u32_t -> uint32_t" epoch.

**Comment by yperess** [comment]
> Ooooph.... standard Grumpy Curmudgeon complaints apply here: touching every file in the project makes patch rebasing across versions a pain, affects users with portability concerns, and generally just leads to headache.
> 
> That said, yes, our include structure is a mess. Yes, it's probably overdue for a cleanup. And yes, I like this scheme better than what we have (rather, "the most charitable interpretation of the design intent of what we have", heh). So I guess this is +1 from me.
> 
> But can we please make sure this is what we really want and that we're sure we're not just going to need to do this again next bikeshed season? I still bear scars from our "int -> u32_t -> uint32_t" epoch.

100% with you, I think that once I get this working fully and passing the tests it'll be easy to tweak. So if you have any preferenmces please let me know.

**Comment by yperess** [comment]
@mbolivar-nordic can we remove the DNM label now that the TSC voted?

**Comment by gmarull** [comment]
+1

File: CMakeLists.txt [hunk 1]
```diff
@@ -106,6 +106,7 @@ zephyr_library_named(zephyr)
 
 zephyr_include_directories(
   include
+  include/zephyr
```
Comment (fabiobaltieri) [inline CMakeLists.txt:109]: Should this be conditional to a Kconfig option? That way we can leave a compatibility mode, but also have an option to help migrate downstream projects or ensure that we don't reintroduce old path includes after this.

Comment (gmarull) [reply CMakeLists.txt:109]: I agree we should have a Kconfig option. Other things I'd consider, given that a future removal of `include/` option will break the compilation of every single file:

- Deprecate current `include/` path, but keep it the default except for upstream code
- Add a big warning when building with the "compatible" mode until it is not deleted (users will notice this way)
- Maybe provide a migration script that automatically prefixes every Zephyr header used in apps/libs with `zephyr/`

Comment (yperess) [reply CMakeLists.txt:109]: Yes, I just wanted to get the CI to pass first before I add another layer to it.

**Comment by yperess** [comment]
I'm not quite sure why the build shows up as broken, locally I'm able to pass: `$ ./scripts/twister -v -i -p nucleo_f429zi -s samples/posix/gettimeofday/sample.posix.gettimeofday`

**Comment by stephanosio** [comment]
> I'm not quite sure why the build shows up as broken, locally I'm able to pass: `$ ./scripts/twister -v -i -p nucleo_f429zi -s samples/posix/gettimeofday/sample.posix.gettimeofday`

@yperess Have you tried `west update`? I am able to reproduce this issue on the latest main.

FYI @erwango @FRASTM

**Comment by yperess** [comment]
OK, so currently everything passes except for the issues in https://github.com/zephyrproject-rtos/zephyr/issues/44486. I'll continue to rebase 2x a day to keep this PR relevant.

**Comment by yperess** [comment]
@mbolivar-nordic can we remove the DNM label on this PR since the TSC voted that we can move forward with review for this change?

**Review by andyross — DISMISSED** [review]
Refresh +1.  Obviously this is way too large to actually review in an exhaustive way manually, but the chosen include scheme is pleasing to my eyes and I can't find any mistakes in the execution.

**Review by cvinayak — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]

**Comment by yperess** [comment]
Rebased (no changes) @povergoing @cvinayak @andyross GitHub automatically dismisses the reviews, could you PTAL? Thanks, and I'm sorry for the inconvinience, hopefully we can merge this quickly.

**Review by jhedberg — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]

**Review by nordic-krch — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
@erwango The failing CI may be of interest to you. It looks like it is related to the STM32 HAL.

**Comment by erwango** [comment]
> @erwango The failing CI may be of interest to you. It looks like it is related to the STM32 HAL.

Fix for this issue just got merged. Can you rebase on latest main ?

**Comment by yperess** [comment]
Rebased, @nordic-krch @povergoing @jhedberg I'm sorry, I need another review (hopefully this can merge before another conflict)

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by carlescufi** [comment]
@yperess one more time

**Comment by nashif** [comment]
> /kernel/include -> /kernel/include/zephyr/kernel/

why is this needed given that kerne/include is internal?

**Review by carlescufi — DISMISSED** [review]

**Review by andyross — DISMISSED** [review]
Refresh +1

**Review by nashif — CHANGES_REQUESTED** [review]
/kernel/include -> /kernel/include/zephyr/kernel/

why is this needed given that kerne/include is internal?

**Comment by yperess** [comment]
> /kernel/include -> /kernel/include/zephyr/kernel/
> 
> why is this needed given that kerne/include is internal?

My thinking was to de-risk. While others may not `#include <zephyr/kernel/mmu.h>` there's nothing stopping an application from using another third party library that also has `mmu.h`. In that case (depending on how imports were done) even the internal build of Zephyr might confuse the Zephyr version of `mmu.h` and the other third party version. Also, it adds consistency.

**Comment by stephanosio** [comment]
>  there's nothing stopping an application from using another third party library that also has mmu.h.

so there are two cases:

1. application code including the header files under `kernel/include`
2. kernel code including header files from the application(s)

The 1st case is not possible since `kernel/include` is only visible to the source files under `kernel/` and `arch/`.
The 2nd case should not be possible either since applications should limit the scope of their include exports to the application target(s).

In that sense, I agree that the `kernel/include` change should not be done.

**Comment by nashif** [comment]
> My thinking was to de-risk. While others may not `#include <zephyr/kernel/mmu.h>` there's nothing stopping an application from using another third party library that also has `mmu.h`. In that case (depending on how imports were done) even the internal build of Zephyr might confuse the Zephyr version of `mmu.h` and the other third party version. Also, it adds consistency.

being able to  include <zephyr/kernel/mmu.h>  and being on the level and under the same namespace makes it look like this header is public and can be used by anyone.

**Review by stephanosio — DISMISSED** [review]

**Comment by yperess** [comment]
Looks like I missed a file at `include/zephyr/types.h` pushed again with that move

**Review by stephanosio — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]
+1 refresh

**Review by nashif — DISMISSED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**